qiskit/transpiler/passmanager.py

```diff
@@ -389,67 +389,63 @@ def __init__(self, stages=None, **kwargs):
             AttributeError: If a stage in the input keyword arguments is not defined.
             ValueError: If an invalid stage name is specified.
         """
-        if stages is None:
-            self.stages = [
-                "init",
-                "layout",
-                "routing",
-                "translation",
-                "optimization",
-                "scheduling",
-            ]
-        else:
-            invalid_stages = [
-                stage for stage in stages if self.invalid_stage_regex.search(stage) is not None
-            ]
-            if invalid_stages:
-                with io.StringIO() as msg:
-                    msg.write(f"The following stage names are not valid: {invalid_stages[0]}")
-                    for invalid_stage in invalid_stages[1:]:
-                        msg.write(f", {invalid_stage}")
-                    raise ValueError(msg.getvalue())
-
-            self.stages = stages
+        stages = stages or [
+            "init",
+            "layout",
+            "routing",
+            "translation",
+            "optimization",
+            "scheduling",
+        ]
+        self._validate_stages(stages)
+        self._stages = stages
         super().__init__()
-        for stage in self.stages:
-            pre_stage = "pre_" + stage
-            post_stage = "post_" + stage
-            setattr(self, pre_stage, None)
+        for stage in self.expanded_stages:
             setattr(self, stage, None)
-            setattr(self, post_stage, None)
         for stage, pm in kwargs.items():
-            if (
-                stage not in self.stages
-                and not (stage.startswith("pre_") and stage[4:] in self.stages)
-                and not (stage.startswith("post_") and stage[5:] in self.stages)
-            ):
+            if stage not in self.expanded_stages:
                 raise AttributeError(f"{stage} is not a valid stage.")
             setattr(self, stage, pm)
```

Comment:
Would it make sense to simplify this to:
```python
for stage in self.expanded_stages:
    pm = kwargs.get(stage, None)
    setattr(self, stage, pm)
```

No `AttributeError` is thrown this way @mtreinish

Reply:
Alternatively:
```python
for stage in self.expanded_stages:
    pm = kwargs.get(stage, None)
    setattr(self, stage, pm)
for stage in kwargs.keys():
    if stage not in self.expanded_stages:
        raise AttributeError(f"{stage} is not a valid stage.")
```
This way we get the `AttributeError` while calling `update_passmanager` only once per stage (+ pre + post)